### ISSUE TITLE
For EventMatch, namedCaptures.length != argumentMetadata.length

### ISSUE THREAD

**Issue by kwalcock** [lum-ai/odinson#186]
For the branch event_match_bug (https://github.com/lum-ai/odinson/tree/event_match_bug) which comes from state_issue, the argumentMetadata isn't synchronizing properly with the namedCaptures.  Their counts are off and their orders are mixed up.  Set a breakpoint at line 49 of OdinsonMatch to watch it happen.

**Comment by BeckySharp** [comment]
good find, thanks!

**Comment by BeckySharp** [comment]
per clarification about the nature of the ArgMetadata aren't parallel lists, and can be matched by `name` attribute.
Maybe it would be easier to retrieve relevant metadata (to check for promotion) if this were a map instead:
https://github.com/lum-ai/odinson/blob/2e60ce6fb215160d69dee46f74a9b7b81d650e34/core/src/main/scala/ai/lum/odinson/OdinsonMatch.scala#L81

**Comment by kwalcock** [comment]
I'm not sure they can be matched by name attribute.  In the array of namedCaptures, there can apparently be multiple with the same name.  For each name, there can be multiple ArgumentMetadata objects with that name.  It doesn't seem possible to know which goes with which even if the order is preserved.  Perhaps I have misunderstood.  I can make it so that if anything with that name should go to state, they all do, or the first one with that name does.  The tests pass that way, but it's probably not right.

**Comment by BeckySharp** [comment]
I think this is what @marcovzla mentioned.  For example, if there is an argument (`name = EXAMPLE`) with a `+` that can be found (and included) one or more times, you may have, say, 3 arguments, each with `name = EXAMPLE`, and only one metadata.  Since it's a map, I think you can iterate through the args, and do something like:
```
for (arg <- arguments) {
  if (metadata(arg.name).promote) {
       // promote(arg) 
  }
}

**Comment by kwalcock** [comment]
I believe to have seen cases in which metadata(arg.name) gets more than one answer.

**Comment by BeckySharp** [comment]
Can you find an example of this?  It may be "by design" and I'm not fully
understanding the design, or it may be a bug

**Comment by kwalcock** [comment]
One of the early examples was this:

I'm seeing EventMatches with two captures and three metadata and their orders also being scrambled. For one in EventTest there are captures for "object" and "subject" and then metadata for "subject", "object", "object".

**Comment by kwalcock** [comment]
Part of the complication is that the NamedCaptures are part of a recursive data structure and for all I know some buried NamedCapture has its promotion turned on.

**Comment by BeckySharp** [comment]
> Part of the complication is that the NamedCaptures are part of a recursive data structure and for all I know some buried NamedCapture has its promotion turned on.

I think the parser syntax wouldn't support that, only the outer argument works with the `^` promote thing.

Do you know which test in TestEvents had the 
> captures for "object" and "subject" and then metadata for "subject", "object", "object".

**Comment by kwalcock** [comment]
It seems to happen for `it should "respect quantifiers in arguments" in {` in TestEvents.scala.

**Comment by BeckySharp** [comment]
thanks! that will help me dig!! more "soon" (tomorrow likely)

**Comment by kwalcock** [comment]
Any luck?  When we discussed it at a meeting, I detected Marco having doubts because of this potential one name to many metadatas possibility.  Can the metadata, at least the one little part I'm concerned with, have different values for the same name?

**Comment by BeckySharp** [comment]
I think this is resolved with the code added in #193, but check it out and let me know if I'm wrong
